strapi/strapi
# fix(database): drop fk before dropping index

### What does it do?

Changes the order of operations in `syncSchema`. In some databases (like mysql) it is impossible to drop an index on a foreign key unless you first drop the fk. The previous order of removal was: index -> fk -> column, I've changed it to fk -> column -> index

### Why is it needed?

see above

more specifically, it also prevents the v4->v5 migration on mysql, because the long version of shortened db identifiers for fks cannot be dropped

### How to test it?

NOTE: there's currently an issue with document id migrations on mysql as well preventing startup, which will prevent you testing this as a v4->v5 migration directly, but otherwise doing a v4->v5 migrations would be the best way to test this since that's what this PR is really aiming to solve

- Using mysql 8
- create your own fk index in the db that strapi won't recognize and will try to delete on startup (or try to migrate from Strapi 4 to Strapi 5)
- start strapi
  - without this PR, it will fail to start because of the error described in https://github.com/strapi/strapi/issues/20020
  - with this PR, the fk will be dropped and Strapi will start fine

Given the comments below, also check some other cases that might trigger a reverse problem. Try restarting strapi after the following cases:

- remove relation (foreign key) from a schema
- add foreign keys to a schema
- rename a relation in a schema
- change the type of a relation to something else while keeping the same name
- swap the names of two attributes, relation and non-relation (but if there's a problem confirm if it exists on v5/main or develop as well, I think there are already potentially some bugs around that area)


Manually tested:

- v4 -> v5 migration
- dropping relations in the CTB
- dropping components in the CTB
- editing relations in the CTB

with

- [x] mariadb

### Related issue(s)/PR(s)

Fixes https://github.com/strapi/strapi/issues/20020

DX-1374
